File: php/api/cluster.php
<?php
/**
 * Cluster API Endpoint
 * Provides cluster-specific information
 *
 * @api
 */

declare(strict_types=1);

header('Content-Type: application/json; charset=UTF-8');
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: GET');

require_once __DIR__ . '/../lib/Storage.php';

try {
    $storage = StorageFactory::createFromEnv();

    // Get cluster name from query parameter
    $clusterName = $_GET['name'] ?? null;
    $type = $_GET['type'] ?? 'overview';

    if (!$clusterName) {
        http_response_code(400);
        echo json_encode([
            'error' => 'Bad Request',
            'message' => 'Cluster name is required'
        ]);
        exit;
    }

    // Sanitize cluster name
    $clusterName = preg_replace('/[^a-zA-Z0-9_-]/', '', $clusterName);

    $response = match($type) {
        'overview' => getClusterOverview($storage, $clusterName),
        'users' => getClusterUsers($storage, $clusterName),
        'disk' => getClusterDisk($storage, $clusterName),
        'history' => getClusterHistory($storage, $clusterName),
        default => ['error' => 'Invalid type parameter']
    };

    // Add metadata
    $response['meta'] = [
        'timestamp' => date('c'),
        'cluster' => $clusterName,
        'type' => $type
    ];

    // Add HATEOAS links
    $response['links'] = [
        'self' => "/api/cluster.php?name={$clusterName}&type={$type}",
        'overview' => "/api/cluster.php?name={$clusterName}&type=overview",
        'users' => "/api/cluster.php?name={$clusterName}&type=users",
        'disk' => "/api/cluster.php?name={$clusterName}&type=disk",
        'history' => "/api/cluster.php?name={$clusterName}&type=history"
    ];

    echo json_encode($response, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE);

} catch (Exception $e) {
    http_response_code(500);
    echo json_encode([
        'error' => 'Internal server error',
        'message' => $e->getMessage()
    ]);
}

/**
 * Get cluster overview
 *
 * @param StorageInterface $storage
 * @param string $clusterName
 * @return array<string, mixed>
 */
function getClusterOverview(StorageInterface $storage, string $clusterName): array
{
    $loadData = $storage->get('load_average') ?? [];
    $pbsData = $storage->get('pbs_usage') ?? [];
    $cpuData = $storage->get('cpu_usage') ?? [];

    $clusterLoad = array_values(array_filter($loadData, fn($item) => $item['cluster'] === $clusterName));
    $clusterPbs = array_values(array_filter($pbsData, fn($item) => $item['cluster'] === $clusterName));
    $clusterCpu = array_values(array_filter($cpuData, fn($item) => $item['cluster'] === $clusterName));

    $hasData = (count($clusterLoad) > 0) || (count($clusterPbs) > 0) || (count($clusterCpu) > 0);

    return [
        'name' => $clusterName,
        'load_average' => (count($clusterLoad) > 0) ? $clusterLoad[0]['value'] : 0,
        'pbs_usage' => (count($clusterPbs) > 0) ? $clusterPbs[0]['value'] : 0,
        'cpu_usage' => (count($clusterCpu) > 0) ? $clusterCpu[0]['value'] : '0/0',
        'has_data' => $hasData,
        'message' => $hasData ? 'OK' : 'データが取得できていません'
    ];
}

/**
 * Get cluster user breakdown
 *
 * @param StorageInterface $storage
 * @param string $clusterName
 * @return array<string, mixed>
 */
function getClusterUsers(StorageInterface $storage, string $clusterName): array
{
    $userData = $storage->get("users_{$clusterName}");

    if ($userData === null || count($userData) === 0) {
        // Generate dummy data for demonstration
        $userData = generateDummyUserData($clusterName);
        $hasData = false;
    } else {
        $hasData = true;
    }

    return [
        'users' => $userData,
        'has_data' => $hasData,
        'message' => $hasData ? 'OK' : 'ダミーデータを表示しています'
    ];
}

/**
 * Get cluster disk usage
 *
 * @param StorageInterface $storage
 * @param string $clusterName
 * @return array<string, mixed>
 */
function getClusterDisk(StorageInterface $storage, string $clusterName): array
{
    $diskData = $storage->get("disk_{$clusterName}");

    if ($diskData === null || count($diskData) === 0) {
        // Generate dummy data for demonstration
        $diskData = generateDummyDiskData($clusterName);
        $hasData = false;
    } else {
        $hasData = true;
    }

    return [
        'nodes' => $diskData,
        'has_data' => $hasData,
        'message' => $hasData ? 'OK' : 'ダミーデータを表示しています'
    ];
}

/**
 * Get cluster history
 *
 * @param StorageInterface $storage
 * @param string $clusterName
 * @return array<string, mixed>
 */
function getClusterHistory(StorageInterface $storage, string $clusterName): array
{
    $days = isset($_GET['days']) ? (int)$_GET['days'] : 7;
    $days = max(1, min($days, 90)); // Limit between 1 and 90 days

    $historyData = $storage->get("history_{$clusterName}_{$days}d");

    if ($historyData === null || count($historyData) === 0) {
        return [
            'history' => [],
            'days' => $days,
            'has_data' => false,
            'message' => '履歴データが見つかりません'
        ];
    }

    return [
        'history' => $historyData,
        'days' => $days,
        'has_data' => true,
        'message' => 'OK'
    ];
}

/**
 * Generate dummy user data
 *
 * @param string $clusterName
 * @return array<int, array<string, mixed>>
 */
function generateDummyUserData(string $clusterName): array
{
    $users = ['alice', 'bob', 'charlie', 'david', 'eve', 'frank', 'grace', 'henry'];
    $result = [];

    foreach ($users as $user) {
        $result[] = [
            'username' => $user,
            'usage' => round(rand(100, 10000) / 10, 2),
            'jobs' => rand(5, 100),
            'last_job' => date('Y-m-d H:i:s', time() - rand(3600, 86400)),
            'is_dummy' => true
        ];
    }

    return $result;
}

/**
 * Generate dummy disk data
 *
 * @param string $clusterName
 * @return array<int, array<string, mixed>>
 */
function generateDummyDiskData(string $clusterName): array
{
    $result = [];
    $mounts = ['/', '/home', '/scratch', '/data'];

    for ($i = 1; $i <= 8; $i++) {
        $disks = [];
        foreach ($mounts as $mount) {
            $total = rand(500, 2000) * 1024 * 1024 * 1024; // GB in bytes
            $used = rand(100, (int)($total * 0.9));

            $disks[] = [
                'mount' => $mount,
                'total' => $total,
                'used' => $used,
                'available' => $total - $used,
                'usage_percent' => round(($used / $total) * 100, 1)
            ];
        }

        $result[] = [
            'name' => "{$clusterName}" . str_pad((string)$i, 2, '0', STR_PAD_LEFT),
            'disks' => $disks,
            'is_dummy' => true
        ];
    }

    return $result;
}

File: backend/internal/api/handlers/cluster.go
package handlers

import (
	"net/http"

	"github.com/taisei-ito/cluster-status-monitor/internal/storage"
)

// ClusterHandler handles cluster API requests
type ClusterHandler struct {
	storage storage.Storage
}

// NewClusterHandler creates a new cluster handler
func NewClusterHandler(storage storage.Storage) *ClusterHandler {
	return &ClusterHandler{storage: storage}
}

// GetClusterInfo handles GET /api/cluster
func (h *ClusterHandler) GetClusterInfo(w http.ResponseWriter, r *http.Request) {
	clusterName := r.URL.Query().Get("name")
	dataType := r.URL.Query().Get("type")

	if clusterName == "" {
		respondJSON(w, http.StatusBadRequest, map[string]string{
			"error": "Cluster name is required",
		})
		return
	}

	var response interface{}
	var err error

	switch dataType {
	case "users":
		response, err = h.getClusterUsers(clusterName)
	case "disk":
		response, err = h.getClusterDisk(clusterName)
	case "history":
		response, err = h.getClusterHistory(clusterName)
	default:
		response, err = h.getClusterSummary(clusterName)
	}

	if err != nil {
		respondJSON(w, http.StatusInternalServerError, map[string]string{
			"error":   "Internal server error",
			"message": err.Error(),
		})
		return
	}

	respondJSON(w, http.StatusOK, response)
}

// getClusterSummary returns summary information for a cluster
func (h *ClusterHandler) getClusterSummary(clusterName string) (map[string]interface{}, error) {
	loadKey := "cluster_" + clusterName + "_load"
	pbsKey := "cluster_" + clusterName + "_pbs"
	cpuKey := "cluster_" + clusterName + "_cpu"

	loadData, _ := h.storage.Get(loadKey)
	pbsData, _ := h.storage.Get(pbsKey)
	cpuData, _ := h.storage.Get(cpuKey)

	hasData := (loadData != nil && len(loadData) > 0) ||
		(pbsData != nil && len(pbsData) > 0) ||
		(cpuData != nil && len(cpuData) > 0)

	result := map[string]interface{}{
		"cluster":      clusterName,
		"load_average": loadData,
		"pbs_usage":    pbsData,
		"cpu_usage":    cpuData,
		"has_data":     hasData,
	}

	if !hasData {
		result["message"] = "データが取得できていません"
	}

	return result, nil
}

// getClusterUsers returns user information for a cluster
func (h *ClusterHandler) getClusterUsers(clusterName string) (map[string]interface{}, error) {
	key := "cluster_" + clusterName + "_users"
	userData, err := h.storage.Get(key)

	if err != nil || len(userData) == 0 {
		return map[string]interface{}{
			"cluster": clusterName,
			"users":   []interface{}{},
			"message": "ユーザーデータが取得できていません",
		}, nil
	}

	return map[string]interface{}{
		"cluster": clusterName,
		"users":   userData,
	}, nil
}

// getClusterDisk returns disk usage information for a cluster
func (h *ClusterHandler) getClusterDisk(clusterName string) (map[string]interface{}, error) {
	key := "cluster_" + clusterName + "_disk"
	diskData, err := h.storage.Get(key)

	if err != nil || len(diskData) == 0 {
		return map[string]interface{}{
			"cluster": clusterName,
			"disk":    map[string]interface{}{},
			"message": "ディスクデータが取得できていません",
		}, nil
	}

	return map[string]interface{}{
		"cluster": clusterName,
		"disk":    diskData,
	}, nil
}

// getClusterHistory returns historical data for a cluster
func (h *ClusterHandler) getClusterHistory(clusterName string) (map[string]interface{}, error) {
	key := "cluster_" + clusterName + "_history"
	historyData, err := h.storage.Get(key)

	if err != nil || len(historyData) == 0 {
		return map[string]interface{}{
			"cluster": clusterName,
			"history": []interface{}{},
			"message": "履歴データが取得できていません",
		}, nil
	}

	return map[string]interface{}{
		"cluster": clusterName,
		"history": historyData,
	}, nil
}
